hal/crc_api.h

```diff
@@ -0,0 +1,178 @@
+    uint32_t initial_xor;
+    /** Final xor value for the computation. */
+    uint32_t final_xor;
+    /** Reflect bits on input. */
+    bool reflect_in;
+    /** Reflect bits in final result before returning. */
+    bool reflect_out;
+} crc_mbed_config_t;
+
+#ifdef DEVICE_CRC
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * \defgroup hal_crc Hardware CRC
+ *
+ * The Hardware CRC HAL API provides a low-level interface to the Hardware CRC
+ * module of a target platform.
+ *
+ * @{
+ */
+
+/** Determine if the current platform supports hardware CRC for given polynomial
+ *
+ * The purpose of this function is to inform the CRC Platform API whether the
+ * current platform has a hardware CRC module and that it can support the
+ * requested polynomial.
+ *
+ * Supported polynomials are restricted to the named polynomials that can be
+ * constructed in the MbedCRC class, POLY_8BIT_CCITT, POLY_7BIT_SD,
+ * POLY_16BIT_CCITT, POLY_16BIT_IBM, POLY_32BIT_ANSI.
+ *
+ * The current platform must support the given polynomials default parameters
+ * in order to return a true response, these include: reflect in, reflect out,
+ * initial xor, and final xor. For example POLY_32BIT_ANSI requires an initial
+ * and final xor of 0xFFFFFFFF, and reflection of both input and output. If any
+ * of these settings cannot be configured the polynomial is not supported.
+ *
+ * \param polynomial CRC Polynomial. Example polynomial: 0x1021 = x^12+x^5+1
+ *
+ * \return  True if running if the polynomial is supported, false if not.
+ */
+bool hal_crc_is_supported(const crc_mbed_config_t* polynomial);
+
+/** Initialise the hardware CRC module with the given polynomial
+ *
+ * After calling this function the CRC HAL module will be ready to receive data
+ * using the hal_crc_compute_partial() function. The CRC module on the board
+ * will be configured internally with the specified configuration and be ready
+ * to receive data.
+ *
+ * The platform will configure itself based on the default configuration
+ * parameters of the input polynomial.
+ *
+ * This function must be called before calling hal_crc_compute_partial().
+ *
+ * This function must be called with a valid polynomial supported by the
+ * platform. The polynomial must be checked for support using the
+ * hal_crc_is_supported() function.
+ *
+ * Calling hal_crc_compute_partial_start() multiple times without finalising the
+ * CRC calculation with hal_crc_get_result() will override the current
+ * configuration and state and the intermediate result of the computation will
+ * be lost.
+ *
+ * This function is not thread safe, a CRC calculation must not be started from
+ * two different threads or contexts at the same time, calling this function
+ * from two different contexts may lead to configurations being overwrite and
+ * results being lost.
+ *
+ * \param polynomial CRC Polynomial. Example polynomial: 0x1021 = x^12+x^5+1
+ */
+void hal_crc_compute_partial_start(const crc_mbed_config_t* polynomial);
```

Comment:
HAL API has first parameter this pointer (*obj), not needed for this API? This is hardware related functionality, thus will be needed to have access to their SDK or internal registers (K64F below has CRC0).

Reply:
I'm not sure I understand. The K64F implementation uses CRC0 as the base pointer for the CRC module because all the CRC functions use offsets from that pointer for accessing specific CRC registers, but not all platforms require anything like that.

Reply:
The point was that all HAL API have defined as first parameter "this" pointer. My question is - why these new functions do not have it

Reply:
We don't need it, there is no platforms with multiple CRC engines (no real need for that) and there's no state/context that we need to be maintained.

Reply:
Some HW CRC modules do store intermediate CRC state, which might be an issue when using CRC HW from different threads.

Reply:
I don't think we want to support multiple hardware CRC calculations being run concurrently, a different thread shouldn't initialize/use the module until the final value has been read from it.

Reply:
Driver is providing the locking, HAL specification explicitly mentions that it's up to the user to ensure access safety. HAL is designed for simplicity, the drivers add the level of complexity necessary for safe usage for the user.